rswag/rswag
# [#443] Update ResponseValidator to support nullable and required

## Problem

Currently running into issues similar to what is described in [this issue](https://github.com/rswag/rswag/issues/443). When I include `nullable` or `required` for response headers it errors out because the response validator checks for `.nil?`.

## Solution

In this PR I add changes to `validate_headers!` checking for `required` and `nullable`. And changing the error return appropriately. I also added tests as well as a subsection to the readme. 

Notably when I ran `./ci/build.sh` and `./ci/test.sh` everything passes, but some of the generated files changed a bit. Happy to change accordingly if there are any recommendations. 

### This concerns this parts of the Open API Specification:
* [OpenAPI Specs – Describing Parameters (section on nullable + required)](https://swagger.io/docs/specification/describing-parameters/)
* [OpenAPI Specs – Describing Responses (section on headers)](https://swagger.io/docs/specification/describing-responses/)

### The changes I made are compatible with:
- [x] OAS2 – I didn't see anything about nullable in OAS2, but this change shouldn't impact current OAS2 users

### Related Issues

Ability to make response headers not mandatory: https://github.com/rswag/rswag/issues/443

### Checklist

### Steps to Test or Reproduce

Can see the line causing the issue [here](https://github.com/rswag/rswag/blob/6f736d8fbfca77a0df7ac0e2b7ef43c09f55df4f/rswag-specs/lib/rswag/specs/response_validator.rb#L37).

When I run `./ci/test.sh` everything works

To replicate can add a nullable and optional header response [here](https://github.com/rswag/rswag/blob/6f736d8fbfca77a0df7ac0e2b7ef43c09f55df4f/rswag-specs/spec/rswag/specs/response_validator_spec.rb#L21).
